File: problem2.cpp
//Time complexity: O(logn)
//Space complexity: O(1)
class Solution {
public:
    int findMin(vector<int>& nums) {
        if(nums.size()==0)
            return -1;
        int l=0, r=nums.size()-1;
        while(l<=r)
        {
            if(nums[l]<nums[r])
                return nums[l];
            int mid=l+(r-l)/2;
            if((mid==0 || nums[mid]<nums[mid-1]) and (nums.size()-1==mid || nums[mid]<nums[mid+1]))
            {
                return nums[mid];
            }
            if(nums[l]<=nums[mid])
                l=mid+1;
            else
                r=mid-1;
        }
        return -1;
    }
};

File: Problem2.py
# Time Complexity : O(logn)
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

# Your code here along with comments explaining your approach
from typing import List
class Solution:
    def findMin(self, nums: List[int]) -> int:
        low = 0
        high = len(nums)-1

        while(low<=high):
            if(nums[low] <= nums[high]): return nums[low]

            mid = low + (high-low)//2

            if(mid != 0 and nums[mid] < nums[mid-1]):
                return nums[mid]
            elif(nums[low]<=nums[mid]):
                low = mid +1
            else:
                high = mid -1
    
        return 99

sc = Solution()
nums = [3,4,5,1,2]
print(sc.findMin(nums))


        